homeassistant/components/moonraker/config_flow.py

```diff
@@ -0,0 +1,212 @@
+        }
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            data = self.discovery_schema or _schema_with_defaults()
+            return self.async_show_form(step_id="user", data_schema=data)
+
+        if user_input[CONF_HOST]:
+            if (
+                not self._reauth_entry
+                and user_input[CONF_HOST] in self._async_current_hosts()
+            ):
+                return self.async_abort(reason="already_configured")
+
+            errors = {}
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_api_key"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                if self._reauth_entry:
+                    self.hass.config_entries.async_update_entry(
+                        self._reauth_entry,
+                        data=user_input,
+                    )
+                    return self.async_abort(reason="reauth_successful")
+                await self.async_set_unique_id(info.get("unique_id"))
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(title=info["title"], data=user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=_schema_with_defaults(
+                user_input[CONF_HOST],
+                user_input[CONF_PORT],
+                user_input[CONF_SSL],
+                user_input[CONF_API_KEY],
+            ),
+            errors=errors,
+        )
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle discovery flow."""
+        local_name = discovery_info.hostname[:-1]
+        node_name = local_name[: -len(".local")]
+        address = discovery_info.properties.get("address", discovery_info.host)
+
+        await self.async_set_unique_id(node_name)
+        self._abort_if_unique_id_configured(updates={CONF_HOST: discovery_info.host})
+
+        if local_name in self._async_current_hosts():
+            return self.async_abort(reason="already_configured")
+        if address in self._async_current_hosts():
+            return self.async_abort(reason="already_configured")
+
+        node_type = f".{discovery_info.type}"
+        self.context["title_placeholders"] = {
+            CONF_HOST: local_name,
+            CONF_NAME: discovery_info.name[: -len(node_type)],
+        }
+
+        self.discovery_schema = _schema_with_defaults(
+            host=local_name, port=discovery_info.port
+        )
+
+        return await self.async_step_user()
```

Comment:
The problem with this approach, is that this will show a form asking the user for host and port, while we already know these. Discovered data should not be in the forms. We should ask the user the least possible.